contracts/asset-proxy/contracts/src/libs/LibAssetData.sol

```diff
@@ -0,0 +1,412 @@
+    }
+
+    /// @dev Decode ERC-721 asset data from the format described in the
+    ///     AssetProxy contract specification.
+    /// @param assetData AssetProxy-compliant asset data describing an ERC-721
+    ///     asset.
+    /// @return The ERC-721 AssetProxy identifier, the address of the ERC-721
+    ///     contract hosting this asset, and the identifier of the specific
+    ///     token to be traded.
+    function decodeERC721AssetData(bytes memory assetData)
+        public
+        pure
+        returns (
+            bytes4 proxyId,
+            address tokenAddress,
+            uint256 tokenId
+        )
+    {
+        proxyId = LibBytes.readBytes4(assetData, 0);
+
+        require(proxyId == ERC721_PROXY_ID);
+
+        tokenAddress = LibBytes.readAddress(assetData, 16);
+        tokenId = LibBytes.readUint256(assetData, 36);
+    }
+
+    /// @dev Encode ERC-1155 asset data into the format described in the
+    ///     AssetProxy contract specification.
+    /// @param tokenAddress The address of the ERC-1155 contract hosting the
+    ///     token(s) to be traded.
+    /// @param tokenIds The identifiers of the specific tokens to be traded.
+    /// @param tokenValues The amounts of each token to be traded.
+    /// @param callbackData ...
+    /// @return AssetProxy-compliant asset data describing the set of assets.
+    function encodeERC1155AssetData(
+        address tokenAddress,
+        uint256[] memory tokenIds,
+        uint256[] memory tokenValues,
+        bytes memory callbackData
+    )
+        public
+        pure
+        returns (bytes memory assetData)
+    {
+        return abi.encodeWithSelector(ERC1155_PROXY_ID, tokenAddress, tokenIds, tokenValues, callbackData);
+    }
+
+    /// @dev Decode ERC-1155 asset data from the format described in the
+    ///     AssetProxy contract specification.
+    /// @param assetData AssetProxy-compliant asset data describing an ERC-1155
+    ///     set of assets.
+    /// @return The ERC-1155 AssetProxy identifier, the address of the ERC-1155
+    ///     contract hosting the assets, an array of the identifiers of the
+    ///     tokens to be traded, an array of token amounts to be traded, and
+    ///     callback data.  Each element of the arrays corresponds to the
+    ///     same-indexed element of the other array.
+    function decodeERC1155AssetData(bytes memory assetData)
+        public
+        pure
+        returns (
+            bytes4 proxyId,
+            address tokenAddress,
+            uint256[] memory tokenIds,
+            uint256[] memory tokenValues,
+            bytes memory callbackData
+        )
+    {
+        proxyId = LibBytes.readBytes4(assetData, 0);
+
+        require(proxyId == ERC1155_PROXY_ID);
+
+        (tokenAddress, tokenIds, tokenValues, callbackData) = abi.decode( // solhint-disable-line indent
+            LibBytes.slice(assetData, 4, assetData.length), (address, uint256[], uint256[], bytes)
```

Comment:
`slice()` has a lot of overhead. `sliceDestructive()` is probably a better choice here, unless side effect is an issue.

I think you can actually just use assembly to extract these values out by pointing the return variables to values/locations in `assetData`. Ping me if you need help.

Reply:
For clarity:

```solidity
assembly {
    // Skip selector and length to get to the first parameter.
    assetData := add(assetData, 36)
    // Read the value of the first parameter.
    tokenAddress := mload(assetData)
    // Point to the next parameter data.
    tokenIds := add(assetData, mload(add(assetData, 32)))
    // Point to the next parameter data.
    tokenvalues := add(assetdata, mload(add(assetdata, 64)))
    // Point to the next parameter data.
    callbackData := add(assetData, add(mload(assetData), 96)))
}
```

Reply:
I'd prefer to minimize the use of assembly and prioritize readability for this contract, since it will probably just be used for calls where efficiency isn't a consideration. I don't mind using `sliceDestructive` here though.

Reply:
Actually, thinking about it more, I could see this being inherited and used internally within other contracts pretty frequently. In that case I'd rather not destroy memory, but efficiency does become more of a factor.